Add lightbox support for photos plugin

iranzo opened the pull request:
Closes #350 

-----
[View rendered CHANGELOG.md](https://github.com/iranzo/elegant/blob/lightbox/CHANGELOG.md)

iranzo commented:
@talha131 I would need your advice on the best way to get this fitting properly with Elegant, as it is works on my site:

![imagen](https://user-images.githubusercontent.com/312463/59271927-24432600-8c55-11e9-83f0-18296474780f.png)

Not sure if there are some other guidelines to follow here

talha131 commented on CHANGELOG.md:
```diff
@@ -1,8 +1,9 @@
 To Be Released
 ===============
 
+* Updated photos plugin integration to show lightbox
```
This line can be removed. B/c we haven't done a release of photos plugin support. So our next release is going to have photos+lightbox

iranzo replied:
I've updated the line to contain both

iranzo replied:
About being optional, we can define a new variable that loads or not the new sections added in the base.html would that be ok?

talha131 replied:
See my new review https://github.com/Pelican-Elegant/elegant/pull/351#discussion_r292445642

talha131 commented:
> I would need your advice on the best way to get this fitting properly with Elegant, as it is works on my site:

Can you please be a bit more specific?

Are you referring to Hound report?

iranzo commented:
No @talha131 I meant in regards of proper location inside the template file, etc.

Hound is now fixed in the same way we fixed 'tipue', ignoring the folder of the library we're not developing

talha131 commented on templates/base.html:
```diff
@@ -49,6 +49,10 @@
         {% include '_includes/feeds.html' %}
         {% endblock feed_links %}
         {% include '_includes/analytics.html' %}
+        {% if (article and article.photo_gallery) %}
```
Why are we adding this in base.html? this plugin depends on `article` variable, right?

Then all the scripts and CSS related to `article` should go into `article.html` template.

iranzo replied:
According  to https://github.com/getpelican/pelican-plugins/tree/master/photos it instructs to place it in 'base'

Initially the loop is for this:

{% if (article and article.photo_gallery) or (articles_page and articles_page.object_list[0].photo_gallery) %}

But I was getting errors on the articles_page.object_list so initially disabled it any idea here?

talha131 replied:
It should work even if you put it in the head of article template. The only requirement is to put it in the head. We already do it for title and meta tags. 

`articles_page` this is for pagination. We don't need it.

Please move into article.html.

iranzo replied:
Seems that the problem is that the areas available in article.html don't render 'after' other libraries are loaded, so it loads first this one without jquery so then, it fails.

How can I tell this one to appear just before </head> ?

iranzo replied:
Probably with this we can move it to article.html and get rid of code in base.html so any advice is appreciated here, thanks

talha131 commented on templates/base.html:
```diff
@@ -102,6 +106,33 @@
+        {% if (article and article.photo_gallery) %}
+        <!-- 'photo' Gallery support -->
+        <!-- Magnific Popup core JS file -->
+        <script src="{{ SITEURL }}/{{ THEME_STATIC_DIR }}/magnific/magnific-popup.js"></script>
```
lightbox support should be optional. We can enable it by default.

1. variable in `_defaults.html` set to ture
1. If it is true then show include`lightbox` script and js
1. If is not true then show vanilla gallery from your previous PR

iranzo replied:
solved in latest push

talha131 commented:
> No @talha131 I meant in regards of proper location inside the template file, etc.

See response https://github.com/Pelican-Elegant/elegant/pull/351#discussion_r292444747

talha131 commented on templates/base.html:
```diff
@@ -102,6 +107,33 @@
         {% endblock script %}
         {% include '_includes/stat_counter.html' %}
+
+        {% if (( PHOTOS_LIGHTBOX == "True" ) and (article and article.photo_gallery)) %}
```
Change it to

```
{% if PHOTOS_LIGHTBOX and article and article.photo_gallery %}
```

You don't need `true` or so many `()`.

iranzo replied:
I did already in my latest code base (maybe yet not pushed)

talha131 commented on templates/base.html:
```diff
@@ -102,6 +107,33 @@
+        <!-- Magnific Popup core JS file -->
+        <script src="{{ SITEURL }}/{{ THEME_STATIC_DIR }}/magnific/magnific-popup.js"></script>
+        <script>
+            $('.gallery').magnificPopup({
```
Just a question. Can we use lightbox without jquery?

iranzo replied:
Quoting: Magnific Popup is a responsive lightbox & dialog script with focus on performance and providing best experience for user with any device
(for jQuery or Zepto.js).

iranzo replied:
So I guess 'no' unless we use zepto.js

iranzo commented on documentation/content/Supported Plugins/photogallery.md:
```diff
@@ -70,6 +70,9 @@ Refer to the plugin documentation for configuration, but at the time of this wri
 `PHOTO_EXIF_COPYRIGHT_AUTHOR = 'Your Name Here'`
 : Adds an author name to the photo's exif and copyright statement. Defaults to `AUTHOR` value from the `pelicanconf.py`
 
+Additionally you can disable 'lightbox' setting (default is enable it):
```
done in next push

talha131 commented on CHANGELOG.md:
```diff
@@ -2,7 +2,7 @@ To Be Released
 
 * New: FontAwesome updated to version 4.7.0
-* New: `Photos` plugin support for photo gallery creation
+* New: `Photos` plugin support for photo gallery creation and lightbox
```
I left a comment earlier. I don't know why github removed it. I said

Release note should read

```
New: `Photos` plugin support for photo gallery creation
New: Lightbox support for `Photos` plugin
```

This will highlight your contribution more.

iranzo replied:
done in next push

talha131 commented on templates/base.html:
```diff
@@ -49,6 +49,11 @@
+        {% from '_includes/_defaults.html' import PHOTOS_LIGHTBOX with context %}
+        {% if PHOTOS_LIGHTBOX and article and article.photo_gallery %}
+        <!-- Gallery css -->
+        <link rel="stylesheet" href="{{ SITEURL }}/{{ THEME_STATIC_DIR }}/magnific/magnific-popup.css">
```
Move it to article.html.

Use `head_links` block.

```
{% block head_links %}
blah blah. Iranzo your code here
{% endblock head_links %}
```

See https://github.com/Pelican-Elegant/elegant/blob/master/templates/article.html#L13-L24 for reference

iranzo replied:
Seems that :

~~~
{% block head_links %}
{% include '_includes/photos_header.html' %}
{% endblock head_links %}
~~~ 

breaks the template and only this css is loaded not others

talha131 commented on templates/base.html:
```diff
@@ -102,6 +107,33 @@
+        <script>
+            $('.gallery').magnificPopup({
+                delegate: 'a',
+                type: 'image',
```
Move this script to article.html `script block`

See https://github.com/Pelican-Elegant/elegant/blob/master/templates/article.html#L119-L123

Also I suggest putting this code in other file and then include it. This will keep the code clean.

iranzo commented on documentation/content/Supported Plugins/photogallery.md:
```diff
@@ -70,6 +70,10 @@ Refer to the plugin documentation for configuration, but at the time of this wri
 `PHOTO_EXIF_COPYRIGHT_AUTHOR = 'Your Name Here'`
 : Adds an author name to the photo's exif and copyright statement. Defaults to `AUTHOR` value from the `pelicanconf.py`
 
+Lightbox is enabled by default. You can disable it with PHOTOS_LIGHTBOX variable. Set it to FALSE in your configuration.
```
done in next push

iranzo commented:
I've uploaded another review which is not working (see my comments above on the header case).

I've also attempted to perform the changes on the documentation file, not sure if is like that what you wanted.

Thanks a lot for your support into getting this merged!

talha131 commented on documentation/content/Supported Plugins/photogallery.md:
````diff
@@ -70,6 +70,9 @@ Refer to the plugin documentation for configuration, but at the time of this wri
 `PHOTO_EXIF_COPYRIGHT_AUTHOR = 'Your Name Here'`
 : Adds an author name to the photo's exif and copyright statement. Defaults to `AUTHOR` value from the `pelicanconf.py`
 
+```PHOTOS_LIGHTBOX = False```
````
Do

<pre>
Lightbox is enabled by default. You can disable it with `PHOTOS_LIGHTBOX` variable. Set it to `FALSE` in your configuration to disable.

```
PHOTOS_LIGHTBOX = False
```

</pre>

Take care of new lines.

iranzo replied:
done, thanks

talha131 commented on templates/article.html:
```diff
@@ -23,6 +23,10 @@
 {{ smo_metadata(article) }}
 {% endblock meta_tags_in_head %}
 
+{% block head_links %}
```
add

```
{ super() }}
```

Immediately after opening the block.

For example https://github.com/Pelican-Elegant/elegant/blob/master/templates/article.html#L17

iranzo replied:
done!, thx

talha131 commented on templates/_includes/photos_footer.html:
```diff
@@ -0,0 +1,27 @@
+{% from '_includes/_defaults.html' import PHOTOS_LIGHTBOX with context %}
```
Does this work? I have a feeling it won't work.

See how we do it with disqus script

https://github.com/Pelican-Elegant/elegant/blob/master/templates/_includes/disqus_scripts.html
https://github.com/Pelican-Elegant/elegant/blob/master/templates/_includes/comments.html

iranzo replied:
It does with the current approach, rendered html contains the block and works fine

talha131 commented:
Thank you for all your work @iranzo
